# Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# importing data

In [3]:
path= r'E:\Asawer\12.06.2024 Instacart Basket Analysis'
#orders data
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
#products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)


# Finding and addressing mixed type variables in a dataframe
### Because our Instacart data does not contain mixed type variables, we will create a test dataframe for practice 

In [6]:
## Create a test dataframe

In [8]:
df_test = pd.DataFrame()

## Create a mixed type column

In [11]:
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [15]:
## Check for any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if weird:
        print(col)

mix


In [17]:
## Convert data in the column "mix" from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
## check datatype again after converting to string
df_test['mix'].dtype

dtype('O')

In [21]:
## Another way to check data type is:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


In [23]:
# Finding and addressing missing values in a dataframe

In [25]:
## Find missing values using isnull().sum() function(it will return true for missing values and false for non missing values)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [27]:
## Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [29]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## How to Handle missing values:
There are some Ways to deal with missing values:
    1. Imputing the NaN with the mean or median of data if it's numerical variable. 
    2. Creating a new variable that acts like a flag based on the missing value.
    3. Removing or filtering out the missing data.

        ### In our case the NaN are string not numerical, so we only have two options to deal with either removing or filtering it.


In [35]:
## Find shape of df
df_prods.shape

(49693, 5)

In [37]:
## Create new df without the nulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [39]:
## the new df should have 16 less rows
df_prods_clean.shape

(49677, 5)

In [41]:
### Another way is to drop all NaN via the following functions:
df_prods.dropna(inplace = True)
### To drop only the NaNs from a particular column :
df_prods.dropna(subset = [‘product_name’], inplace = True)

SyntaxError: invalid character '‘' (U+2018) (3908009250.py, line 4)

In [43]:
# Finding and addressing duplicate values in a dataframe

In [45]:
## Create subset of df_prods_clean that contains only rows of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [47]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [49]:
## Check current number of rowns in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [51]:
## Create a new datafram with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [53]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting Changes

In [56]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

## Step 2

In [60]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Notes:
1. According to the statisics and starting with count, "days_since_prior_order" has less rows count compared to the others indicating missing data.
2. the minimum of order number= 1 and maximum=100 indicating that it ranges from 1-100. The range of days of weeks is 0-6. For the hour of day, the range is 0-24 hours. data decription showed that days since prior order is limited to 30 days ( 1 month).

## Step 3

In [64]:
## Check for any mixed-type columns
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)

Unnamed: 0
order_id
user_id
order_number
order_dow
order_hour_of_day
days_since_prior_order


In [66]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


## Step 4
no mixed-type data were found.

## Step 5

In [70]:
## Find missing values using isnull().sum() function(it will return true for missing values and false for non missing values)
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### days_since_prior_order has 206209 NaN. this might indicate that most of orders were done at the same day. or some data is not recorded for this column.

## Step 6

In [74]:
### Create subset of the dataframe that contains NaN or missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [76]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [78]:
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [80]:
## Step 7

In [82]:
# Create subset of df_ords that contains only rows of duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [84]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

### Notes:
No duplicates were found. If I found duplicates, I would run '.drop_duplicates()' to creat a new datafreame without duplicates.

In [ ]:
## Step 9

In [85]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))